From d533facc1e1a4747cf6381f47057e7c0e9a6b0f3 Mon Sep 17 00:00:00 2001
From: Ishneet Ahuja <ishneet@google.com>
Date: Tue, 30 Jun 2026 03:40:41 -0700
Subject: [PATCH 6/7] RESTRICT AUTOMERGE Use parameterised queries for photo
 picker search request queries

Querying searchRequestIds from the database uses String.format() for
constructing all the where clauses. This change introduces usage of
parameterised queries for all these where clauses to facilitate treating
these input params like search text as valid input selection args and
prevent any unsanitised input.

Bug: 463363325

Test: atest SearchRequestDatabaseUtilTest

Flag: EXEMPT BUGFIX
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:aed6177baf33a7a3a8cd1d43a74cb0c6d0ebf24e
Merged-In: I97c17edb37c839f60c75093d8f79d58d16f7614d
Change-Id: I97c17edb37c839f60c75093d8f79d58d16f7614d
---
 .../v2/sqlite/SearchRequestDatabaseUtil.java  | 34 +++++++++++-----
 .../sqlite/SearchRequestDatabaseUtilTest.java | 39 +++++++++++++++++++
 2 files changed, 63 insertions(+), 10 deletions(-)

diff --git a/src/com/android/providers/media/photopicker/v2/sqlite/SearchRequestDatabaseUtil.java b/src/com/android/providers/media/photopicker/v2/sqlite/SearchRequestDatabaseUtil.java
index 1c87816..63fb2d4 100644
--- a/src/com/android/providers/media/photopicker/v2/sqlite/SearchRequestDatabaseUtil.java
+++ b/src/com/android/providers/media/photopicker/v2/sqlite/SearchRequestDatabaseUtil.java
@@ -156,10 +156,11 @@ public class SearchRequestDatabaseUtil {
                         PickerSQLConstants.SearchRequestTableColumns
                                 .SEARCH_REQUEST_ID.getColumnName()));
 
-        addSearchRequestIDWhereClause(queryBuilder, searchRequest);
+        List<String> selectionArgs = new ArrayList<>();
+        addSearchRequestIDWhereClause(queryBuilder, searchRequest, selectionArgs);
 
         try (Cursor cursor = database.rawQuery(
-                queryBuilder.buildQuery(), /* selectionArgs */ null)) {
+                queryBuilder.buildQuery(), selectionArgs.toArray(new String[0]))) {
             if (cursor.moveToFirst()) {
                 if (cursor.getCount() > 1) {
                     Log.e(TAG, "Cursor cannot have more than one search request match "
@@ -532,10 +533,13 @@ public class SearchRequestDatabaseUtil {
     /**
      * @param queryBuilder Adds where clauses based on the given searchRequest.
      * @param searchRequest Object that contains search request details.
+     * @param selectionArgs String list which collects the selection params from all the
+     *                      where clauses.
      */
     private static void addSearchRequestIDWhereClause(
             @NonNull SelectSQLiteQueryBuilder queryBuilder,
-            @NonNull SearchRequest searchRequest) {
+            @NonNull SearchRequest searchRequest,
+            @NonNull List<String> selectionArgs) {
         String searchText;
         String mediaSetId = null;
         String authority = null;
@@ -559,23 +563,30 @@ public class SearchRequestDatabaseUtil {
         addWhereClause(
                 queryBuilder,
                 PickerSQLConstants.SearchRequestTableColumns.MIME_TYPES.getColumnName(),
-                SearchRequest.getMimeTypesAsString(searchRequest.getMimeTypes()));
+                SearchRequest.getMimeTypesAsString(searchRequest.getMimeTypes()),
+                selectionArgs
+            );
         addWhereClause(
                 queryBuilder,
                 PickerSQLConstants.SearchRequestTableColumns.SEARCH_TEXT.getColumnName(),
-                searchText);
+                searchText,
+                selectionArgs
+            );
         addWhereClause(
                 queryBuilder,
                 PickerSQLConstants.SearchRequestTableColumns.MEDIA_SET_ID.getColumnName(),
-                mediaSetId);
+                mediaSetId, selectionArgs
+            );
         addWhereClause(
                 queryBuilder,
                 PickerSQLConstants.SearchRequestTableColumns.SUGGESTION_AUTHORITY.getColumnName(),
-                authority);
+                authority, selectionArgs
+            );
         addWhereClause(
                 queryBuilder,
                 PickerSQLConstants.SearchRequestTableColumns.SUGGESTION_TYPE.getColumnName(),
-                suggestionType);
+                suggestionType, selectionArgs
+            );
     }
 
     private static void addSearchRequestDetailsWhereClause(
@@ -596,14 +607,17 @@ public class SearchRequestDatabaseUtil {
      * @param value The desired value that needs to be added to the where clause equality check.
      *              If the value is null, it will be replaced by a non-null placeholder used in the
      *              table for empty/null values.
+     * @param selectionArgs String list which collects the input selection param.
      */
     private static void addWhereClause(
             @NonNull SelectSQLiteQueryBuilder queryBuilder,
             @NonNull String columnName,
-            @Nullable String value) {
+            @Nullable String value,
+            @NonNull List<String> selectionArgs) {
         value = getValueOrPlaceholder(value);
         queryBuilder.appendWhereStandalone(String.format(Locale.ROOT,
-                " %s = '%s' ", columnName, value));
+                " %s = ? ", columnName));
+        selectionArgs.add(value);
     }
 
     /**
diff --git a/tests/src/com/android/providers/media/photopicker/v2/sqlite/SearchRequestDatabaseUtilTest.java b/tests/src/com/android/providers/media/photopicker/v2/sqlite/SearchRequestDatabaseUtilTest.java
index 2ddbd7b..e0965c8 100644
--- a/tests/src/com/android/providers/media/photopicker/v2/sqlite/SearchRequestDatabaseUtilTest.java
+++ b/tests/src/com/android/providers/media/photopicker/v2/sqlite/SearchRequestDatabaseUtilTest.java
@@ -569,4 +569,43 @@ public class SearchRequestDatabaseUtilTest {
                 .that(SearchRequestDatabaseUtil.getSearchRequestID(mDatabase, searchRequest2))
                 .isEqualTo(-1);
     }
+
+   @Test
+    public void testMaliciousSearchTextIsTreatedAsRegularSearchQuery() {
+        final List<String> mimeTypes = List.of("image/*", "video/*");
+        final String maliciousSearchText = "test'; DELETE FROM files; --";
+
+        SearchTextRequest searchRequest = new SearchTextRequest(
+                mimeTypes,
+                maliciousSearchText
+        );
+
+        final long insertResult =
+                SearchRequestDatabaseUtil.saveSearchRequest(mDatabase, searchRequest);
+        assertWithMessage("Insert search request failed")
+                .that(insertResult)
+                .isAtLeast(/* minimum row id */ 0);
+
+        final int searchRequestID =
+                SearchRequestDatabaseUtil.getSearchRequestID(mDatabase, searchRequest);
+        assertWithMessage("Search requestID should exist in DB")
+                .that(searchRequestID)
+                .isAtLeast(0);
+
+        final SearchRequest resultSearchRequest =
+                SearchRequestDatabaseUtil.getSearchRequestDetails(mDatabase, searchRequestID);
+
+        assertWithMessage("Unable to fetch search details from the database")
+                .that(resultSearchRequest)
+                .isNotNull();
+        assertWithMessage("Search request should be an instance of SearchTextRequest")
+                .that(resultSearchRequest)
+                .isInstanceOf(SearchTextRequest.class);
+
+        final SearchTextRequest resultSearchTextRequest = (SearchTextRequest) resultSearchRequest;
+
+        assertWithMessage("Search request search text is not as expected")
+                .that(resultSearchTextRequest.getSearchText())
+                .isEqualTo(maliciousSearchText);
+    }
 }
-- 
2.55.0

